File: client-domain/src/main/java/com/jraska/github/client/DeepLinkHandler.java
package com.jraska.github.client;

import com.jraska.github.client.analytics.AnalyticsEvent;
import com.jraska.github.client.analytics.EventAnalytics;
import com.jraska.github.client.logging.CrashReporter;
import okhttp3.HttpUrl;

import javax.inject.Inject;

public final class DeepLinkHandler {
  private final DeepLinkLauncher linkLauncher;
  private final WebLinkLauncher fallbackLauncher;
  private final CrashReporter crashReporter;
  private final EventAnalytics eventAnalytics;

  @Inject
  public DeepLinkHandler(DeepLinkLauncher linkLauncher, WebLinkLauncher fallbackLauncher,
                         CrashReporter crashReporter, EventAnalytics eventAnalytics) {
    this.linkLauncher = linkLauncher;
    this.fallbackLauncher = fallbackLauncher;
    this.crashReporter = crashReporter;
    this.eventAnalytics = eventAnalytics;
  }

  public void handleDeepLink(HttpUrl deepLink) {
    AnalyticsEvent event = AnalyticsEvent.builder("deep_link_received")
      .addProperty("deep_link", deepLink.toString())
      .build();
    eventAnalytics.report(event);

    try {
      linkLauncher.launch(deepLink);
    } catch (IllegalArgumentException ex) {
      crashReporter.report(ex, "Invalid deep link", deepLink.toString());
      fallbackLauncher.launch(deepLink);
    }
  }
}

File: client-domain/src/main/java/com/jraska/github/client/DeepLinkHandler.kt
package com.jraska.github.client

import com.jraska.github.client.analytics.AnalyticsEvent
import com.jraska.github.client.analytics.EventAnalytics
import com.jraska.github.client.common.toAnalyticsString
import com.jraska.github.client.logging.CrashReporter

import okhttp3.HttpUrl
import javax.inject.Inject

class DeepLinkHandler @Inject constructor(
  private val linkLauncher: DeepLinkLauncher,
  private val fallbackLauncher: WebLinkLauncher,
  private val crashReporter: CrashReporter,
  private val eventAnalytics: EventAnalytics
) {

  fun handleDeepLink(deepLink: HttpUrl) {
    var success = false

    try {
      linkLauncher.launch(deepLink)
      success = true
    } catch (ex: IllegalArgumentException) {
      crashReporter.report(ex, "Invalid deep link", deepLink.toString())
      fallbackLauncher.launch(deepLink)
    }

    val event = AnalyticsEvent.builder("deep_link_received")
      .addProperty("deep_link", deepLink.toAnalyticsString())
      .addProperty("success", success)
      .build()
    eventAnalytics.report(event)
  }
}
